go/vt/vttablet/grpctmclient/client.go

```diff
@@ -152,11 +166,48 @@ func (client *grpcClient) dial(ctx context.Context, tablet *topodatapb.Tablet) (
 	return tabletmanagerservicepb.NewTabletManagerClient(cc), cc, nil
 }
 
-func (client *grpcClient) dialPool(ctx context.Context, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, error) {
+func (client *grpcClient) dialPool(ctx context.Context, dialPoolGroup DialPoolGroup, tablet *topodatapb.Tablet) (tabletmanagerservicepb.TabletManagerClient, invalidatorFunc, error) {
 	addr := netutil.JoinHostPort(tablet.Hostname, int32(tablet.PortMap["grpc"]))
 	opt, err := grpcclient.SecureDialOption(cert, key, ca, crl, name)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
+	}
+
+	createTmc := func() (*tmc, error) {
+		cc, err := grpcclient.Dial(addr, grpcclient.FailFast(false), opt)
+		if err != nil {
+			return nil, err
+		}
+		return &tmc{
+			cc:     cc,
+			client: tabletmanagerservicepb.NewTabletManagerClient(cc),
+		}, nil
+	}
+
+	if dialPoolGroup != DialPoolGroupDefault {
+		client.mu.Lock()
+		defer client.mu.Unlock()
+		if client.rpcDialPoolMap == nil {
+			client.rpcDialPoolMap = make(map[DialPoolGroup]addrTmcMap)
+		}
+		if _, ok := client.rpcDialPoolMap[dialPoolGroup]; !ok {
+			client.rpcDialPoolMap[dialPoolGroup] = make(addrTmcMap)
+		}
+		m := client.rpcDialPoolMap[dialPoolGroup]
+		if _, ok := m[addr]; !ok {
+			tm, err := createTmc()
+			if err != nil {
+				return nil, nil, err
+			}
+			m[addr] = tm
+		}
+		invalidator := func(err error) {
+			client.mu.Lock()
+			defer client.mu.Unlock()
+			m[addr].cc.Close()
+			delete(m, addr)
+		}
```

Comment:
Just one question. We take in the error here, but don't check for the error to be a connection error, or even if it's nil. Won't this cause us to invalidate the connections after every call?

Reply:
we call invalidator only when there is an error, passing of error should removed

Reply:
Good catch on not checking for `err == nil`! Updated.
Re: connection error, we've already established last week that it's very difficult to distinguish between gRPC errors, TCP errors, vitess errors, MySQL errors etc., because the call stack is such that errors are wrapped up as `SQLErr` including gRPC ones. We therefore invalidate on _any_ error.

Mind you that in this PR we only create an invalidation function for `FullStatus` and for `CheckThrottler`, where SQL errors are highly unexpected, as opposed to `ExecuteFetch*` where such errors are trivial.

Reply:
No we don't. At least not in the code right now. We call it without checking for an error.

Reply:
Yep, makes sense!

Reply:
I further updated to reflect @harshit-gangal 's comment. Calling `invalidator()` only on error, and now not even passing the error argument.